src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,505 @@
+        methodobj[jss::duration_us] = std::to_string(
+            std::chrono::duration_cast<microseconds>(
+                present - m.second.second).count());
+        methodsArray.append(methodobj);
+    }
+
+    Json::Value current(Json::objectValue);
+    current[jss::jobs] = jobsArray;
+    current[jss::methods] = methodsArray;
+    return current;
+}
+
+//-----------------------------------------------------------------------------
+
+void
+PerfLogImp::openLog()
+{
+    if (! setup_.perf_log.empty())
+    {
+        if (logFile_.is_open())
+            logFile_.close();
+
+        auto log_dir = setup_.perf_log.parent_path();
+        if (!boost::filesystem::is_directory(log_dir))
+        {
+            boost::system::error_code ec;
+            boost::filesystem::create_directories(log_dir, ec);
+            if (ec)
+            {
+                journal_.fatal() << "Unable to create performance log "
+                    "directory " << log_dir << ": " << ec.message();
+                app_.signalStop();
+                return;
+            }
+        }
+
+        try
+        {
+            logFile_.open(setup_.perf_log.c_str(),
+                std::ios::out | std::ios::app);
+        }
+        catch (std::exception const& e)
+        {
+            journal_.fatal() << "Unable to open performance log " <<
+                setup_.perf_log << ": " << e.what();
+            app_.signalStop();
+        }
+    }
+}
+
+void
+PerfLogImp::run()
+{
+    beast::setCurrentThreadName("perflog");
+    lastLog_ = system_clock::now();
+
+    while (true)
+    {
+        {
+            std::unique_lock<std::mutex> lock(mutex_);
+            if (stop_)
+            {
+                stopped();
+                return;
+            }
+            if (rotate_)
+            {
+                openLog();
+                rotate_ = false;
+            }
+            cond_.wait_until(lock,
+                lastLog_ + setup_.log_interval);
+        }
+        report();
+    }
+}
+
+void
+PerfLogImp::report()
+{
+    auto present = system_clock::now();
+
+    if (present < lastLog_ + setup_.log_interval)
+        return;
+    lastLog_ = present;
+
+    Json::Value report(Json::objectValue);
+    report[jss::time] = to_string(date::floor<microseconds>(present));
+    report[jss::workers] = counters_.workers_;
+    report[jss::hostid] = cHostname_;
+    report[jss::counters] = counters_.countersJson();
+    auto cur = counters_.currentJson();
+    report[jss::current_activities] = counters_.currentJson();
+
+    logFile_ << Json::Compact{std::move(report)} << std::endl;
+}
+
+PerfLogImp::PerfLogImp(Setup const& setup,
+    Stoppable& parent,
+    beast::Journal journal,
+    Application& app)
+        : Stoppable ("PerfLogImp", parent)
+        , setup_ (setup)
+        , journal_ (journal)
+        , app_ (app)
+{
+    openLog();
+    counters_.populateRpc(ripple::RPC::getHandlerNames());
+    counters_.populateJq(JobTypes().typeNames());
```

Comment:
I still have a sense that this is a misuse of `JobTypes`.  Perhaps there's a solution that can make us both happy.  The problem seems to lie in the fact that `JobQueue` believes it is the only thing that needs to use the `JobTypes`, but now `PerfLog` has a legitimate use.

To address this, I suggest that the singleton `JobTypes` currently being held by `JobQueue` be moved out of `JobQueue` to be owned by `JobTypes`.  This change allows both `JobQueue` and `PerfLog` equally convenient access to the `JobTypes`.  Making `JobTypes` a public singleton should not run any thread-related risks since all of its members are `const`.

You can see the change I'm proposing here: https://github.com/scottschurr/rippled/commit/0b7873998b0ee2bd2d9f9ebe43b6456c9cb40b02.